### PR TITLE
Detect snap installation of Visual Studio Code on Linux

### LINKED ISSUE #155: Editors installed via snap are not detected
## Description
I have visual studio code installed:
```bash
➜  ~ whereis code
code: /var/lib/snapd/snap/bin/code /var/lib/snapd/snap/bin/code.url-handler
```
but it's not detected:
Screenshot from 2019-06-18 11-25-47

## Version
* GitHub Desktop: 2.0.4-linux1
* Operating system: Fedora 29

## Steps to Reproduce
1. install [visual studio code via snap](https://snapcraft.io/code) 
2. try to set it as default editor

### Expected Behavior
Visual Studio Code should be detected.

### Actual Behavior
Visual Studio Code is not detected.

I'll look into this in a near future if no one has time to fix it :+1:

### PR DESCRIPTION
**Pull request by ziggy42**
**Closes #155**

## Description

Allows Visual Studio Code and Visual Studio Code Insiders to be detected as external editors when installed as snaps

## Release notes

Notes: detect Visual Studio Code and Visual Studio Code Insiders as external editors when installed as snaps

### CODE REVIEW

**Comment by ziggy42** [comment]
Not sure I understand why the linux build is failing...

File: app/src/lib/editors/linux.ts [hunk 1]
```diff
@@ -44,32 +44,40 @@ async function getPathIfAvailable(path: string): Promise<string | null> {
...
     case ExternalEditor.Atom:
       return getPathIfAvailable('/usr/bin/atom')
     case ExternalEditor.VisualStudioCode:
-      return getPathIfAvailable('/usr/bin/code')
+      return getFirstPathIfAvailable(['/snap/bin/code', '/usr/bin/code'])
     case ExternalEditor.VisualStudioCodeInsiders:
-      return getPathIfAvailable('/usr/bin/code-insiders')
+      return getFirstPathIfAvailable([
+        '/snap/bin/code-insiders',
+        '/usr/bin/code-insiders',
```
Comment (LuKks) [inline app/src/lib/editors/linux.ts]:
```suggestion
        '/snap/bin/code-insiders',
```

File: app/src/lib/editors/linux.ts [hunk 2]
```diff
@@ -44,32 +44,40 @@ async function getPathIfAvailable(path: string): Promise<string | null> {
...
   switch (editor) {
     case ExternalEditor.Atom:
       return getPathIfAvailable('/usr/bin/atom')
     case ExternalEditor.VisualStudioCode:
-      return getPathIfAvailable('/usr/bin/code')
+      return getFirstPathIfAvailable(['/snap/bin/code', '/usr/bin/code'])
     case ExternalEditor.VisualStudioCodeInsiders:
-      return getPathIfAvailable('/usr/bin/code-insiders')
+      return getFirstPathIfAvailable([
+        '/snap/bin/code-insiders',
```
Comment (LuKks) [inline app/src/lib/editors/linux.ts]:
```suggestion
        '/usr/bin/code-insiders',
```

File: app/src/lib/editors/linux.ts [hunk 3]
```diff
@@ -44,32 +44,40 @@ async function getPathIfAvailable(path: string): Promise<string | null> {
...
+  return null
+}
+
 async function getEditorPath(editor: ExternalEditor): Promise<string | null> {
   switch (editor) {
     case ExternalEditor.Atom:
       return getPathIfAvailable('/usr/bin/atom')
     case ExternalEditor.VisualStudioCode:
-      return getPathIfAvailable('/usr/bin/code')
+      return getFirstPathIfAvailable(['/snap/bin/code', '/usr/bin/code'])
```
Comment (LuKks) [inline app/src/lib/editors/linux.ts]:
```suggestion
      return getFirstPathIfAvailable(['/usr/bin/code', '/snap/bin/code'])
```

Comment (LuKks) [inline app/src/lib/editors/linux.ts:64]:
```suggestion
      return getFirstPathIfAvailable(['/usr/bin/code', '/snap/bin/code'])
```

File: app/src/lib/editors/linux.ts (hunk 2 above)
Comment (LuKks) [inline app/src/lib/editors/linux.ts:67]:
```suggestion
        '/usr/bin/code-insiders',
```

File: app/src/lib/editors/linux.ts (hunk 1 above)
Comment (LuKks) [inline app/src/lib/editors/linux.ts:68]:
```suggestion
        '/snap/bin/code-insiders',
```

**Comment by LuKks** [comment]
Can you try that? Just swapping the directories. I don't know how it works but it's the only thing that seems affecting the process and not being backward compatible.

edit: except if you really need that match first with snap, in that case my changes are useless.

Also you check this out?
https://brendanforster.visualstudio.com/desktop-linux/_build/results?buildId=385
Linux -> Package in Container

**Comment by shiftkey** [comment]
> Not sure I understand why the linux build is failing...

@ziggy42 I don't believe it's related to the PR. I'll investigate this when I have a spare moment.

**Comment by shiftkey** [comment]
> Can you try that? Just swapping the directories. I don't know how it works but it's the only thing that seems affecting the process and not being backward compatible.

@LuKks the order of these paths should only matter if you have both installed. Something needs to be chosen, and a pattern that's good to go with here is something like:

 - when multiple versions are found, favour the latest version
 - when multiple installations found of the same version, favour the one that's more specific to the user

I don't have strong feelings on what order we should use here, just that we choose _something_ and try to stick to it...

**Review by shiftkey — APPROVED** [review]

**Pull request merged**